Placeholder implementation

veerajsk opened the pull request:
Design doc: https://docs.google.com/document/d/1XoiiQ_KQrD4j--jwxyoB5GtDZt1-pcBMxRibgyl-Crw/edit#heading=h.g15870th7faw

Note: Frontend & backend do not yet support placeholders therefore plan changes for turtlesim are not shown here

Brief PR overview:
The engine now supports placeholders in ConfAbstractPlanWrapper objects. The placeholders are mapped to their implementations by looking up a mapping json string specified as a parameter to the AlicaContext

As a demonstration, the turtlesim app is updated to use placeholders for the WaitForTrigger & WaitForMsg which have different implementations based on whether ros1 or ros2 is used

veerajsk committed:
Placeholder initial implementation

veerajsk commented:
What's done: a working implementation of placeholders with turtlesim 
What's left: unit tests + minor error checking while parsing the plans + more thorough testing

Dale-Koenig commented on alica_engine/include/engine/model/Placeholder.h:
```diff
@@ -0,0 +1,26 @@
+private:
+    friend PlaceholderFactory;
+    /**
+     * If nullptr, it will simply receive a reference to its parents Blackboard
```
I think we should never allow for placeholders taking by reference, (i.e. placeholders should not allow inheriting blackboard)

veerajsk replied:
This seems like a bug actually in the engine. It does not distinguish between an empty blackboard & inherited blackboard. I'll fix

veerajsk replied:
this was a misunderstanding on my part of the code. Basically the comment here is incorrect, because createEmpty() is used to create an empty blackboard & it will never be null. I updated the comment

Dale-Koenig commented on alica_engine/include/engine/model/PlaceholderMapping.h:
```diff
@@ -0,0 +1,26 @@
+class PlaceholderMappingFactory;
+class AbstractPlan;
+
+class PlaceholderMapping : public AlicaElement
```
This isn't really an alica element right, since it is not created in plan designer?  It is more just a config on par with Alica.yaml file.  In particular I think it may not have id/name

veerajsk replied:
I guess we can look at it either way. In any case the implementation will have to be as if it is an alica element I think because we need this file to parse the plan tree. But sure, externally we can remove id/name. 
One concern if we don't want to set this file through the designer is that we may mess up the ids of the plans/behs & also mess up the blackboard mapping.

veerajsk replied:
Another thing to note is in the current implementation we can map either a behaviour or a plan to a placeholder. This is pretty apparent I guess, but just putting it here so that there is no confusion

Dale-Koenig replied:
Yeah this is intentional (though really now that plans have run() method etc, behaviours are basically just plans with a single entrypoint/state)

veerajsk replied:
I spent quite a bit of time on this but I think from an implementation perspective it is easiest to treat the mapping as an alica element. However, I did change how the mapping is specified as you suggested. Now the engine takes the mapping json directly instead of a file name. Additionally there is no need to specify id, name & comment for the mapping. So basically the mapping is treated as an alica element only within the engine as an implementation convenience

Dale-Koenig commented on alica_engine/include/engine/modelmanagement/factories/PlaceholderMappingFactory.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "Factory.h"
+
+#include <yaml-cpp/yaml.h>
```
nit:  Maybe better to forward declare class and only include in cpp file if possible?

veerajsk replied:
The create function takes a YAML::Node, so we need to include whatever file makes this class visible. I think yaml.h is the only one or I'm not sure, will have to check

Dale-Koenig commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -36,13 +40,22 @@ ModelManager::ModelManager(ConfigChangeListener& configChangeListener, const std
 }
 
-Plan* ModelManager::loadPlanTree(const std::string& masterPlanName)
+Plan* ModelManager::loadPlanTree(const std::string& masterPlanName, std::optional<std::string> placeholderMapping)
```
IMO it's better if we left the file loading to the application and took in a loaded file as a string/etc that can be parsed to yaml.  That way we can potentially even read the full file from e.g. the ROS param server (allowing fussing with placeholders using rosarg).  Requiring the mapping to be written in hardcoded file disallows some customization

veerajsk replied:
sure, can be done

veerajsk replied:
I implemented this

Dale-Koenig commented on supplementary/alica_turtlesim/libalica-turtlesim/src/Teleport.cpp:
```diff
@@ -15,11 +15,25 @@ Teleport::Teleport(alica::BehaviourContext& context)
+        return;
+    }
+
+    // Delay so as to avoid rapid success in case of repeated calls
```
What is the problem here?  setSuccess() should mean repeated calls return immediately right?

veerajsk replied:
In PlaceholderTutorial the MoveSequence plan moves the robot to each of the 4 corners using the IGoTo interface. If Teleport is used here without the delay it will move so fast barely anything is visible. Therefore I put a delay here. I'll improve the comment

Dale-Koenig replied:
Perhaps better is to add time as blackboard input to teleport plan?

veerajsk replied:
no longer relevant since we use placeholders for waitformsg & waitfortrigger behs

ntfshard commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -105,6 +106,7 @@ struct AlicaContextParams
             , configPaths{configPath}
             , roleSetName(roleSetName)
             , masterPlanName(masterPlanName)
+            , placeholderMapping()
```
minor: redundant

veerajsk replied:
yes, but I think its better to list all members here to avoid confusion & future mistakes

ntfshard commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -120,15 +122,19 @@ struct AlicaContextParams
     AlicaContextParams(const std::string& agentName, const std::vector<std::string>& configPaths, const std::string& roleSetName,
-            const std::string& masterPlanName, bool stepEngine = false, const AgentId agentID = InvalidAgentID)
+            const std::string& masterPlanName, bool stepEngine = false, const AgentId agentID = InvalidAgentID,
+            std::optional<std::string> placeholderMapping = std::nullopt)
```
why do we need optional? empty string also can indicate not selected file

ntfshard replied:
Here and around this PR

veerajsk replied:
I generally prefer an optional to having a specific value denote that the value is not specified. I think this is more apparent now that this value represents a json

ntfshard commented on alica_engine/src/engine/modelmanagement/factories/PlaceholderMappingFactory.cpp:
```diff
@@ -0,0 +1,36 @@
+    storeElement(mapping, alica::Strings::placeholderMapping);
+
+    YAML::Node mappingNode = node[alica::Strings::placeholderMapping];
+    for (YAML::Node entry : mappingNode) {
```
do we need copying `entry` here?

veerajsk replied:
I can't seem to find documentation for the YAML node class & it also seems to be a bit too complicated & versatile. So I'm not sure if a copy has overhead or is needed here. Honestly I just looked at code that parses an array & used the same pattern here.

ntfshard commented on alica_engine/src/engine/modelmanagement/factories/PlaceholderMappingFactory.cpp:
```diff
@@ -0,0 +1,36 @@
+
+    YAML::Node mappingNode = node[alica::Strings::placeholderMapping];
+    for (YAML::Node entry : mappingNode) {
+        placeholderReferences.push_back({mapping->getId(), getReferencedId(entry[alica::Strings::placeholder])});
+        placeholderAbstractPlanReferences.push_back({mapping->getId(), getReferencedId(entry[alica::Strings::placeholderImplementation])});
```
we can reserve space before loop

veerajsk replied:
same as earlier comment. I'm having a hard time figuring out the Node class to be able to utilize all its functionality.

bjoernschroeder replied:
placeholderReferences / placeholderAbstractPlanReferences are lists, not vectors

ntfshard commented on alica_engine/src/engine/modelmanagement/factories/PlaceholderMappingFactory.cpp:
```diff
@@ -0,0 +1,36 @@
+            placeholderIt != placeholderReferences.end(); ++placeholderIt, ++abstractPlanIt) {
+        assert(placeholderIt->first == abstractPlanIt->first);
+        auto mapping = dynamic_cast<PlaceholderMapping*>(const_cast<AlicaElement*>(getElement(placeholderIt->first)));
+        // TODO: verify that the placeholder & abstract plans are compatible
```
don't want to add assert against nullptr here?

ntfshard commented on alica_engine/src/engine/model/Placeholder.cpp:
```diff
@@ -0,0 +1 @@
+#include <engine/model/Placeholder.h>
```
short file. why?

veerajsk replied:
so that the header is compiled

ntfshard replied:
This header already included in other places.
If you insists, let it be

veerajsk replied:
linking was failing if I didn't have this file, therefore I added it.

ntfshard commented on alica_engine/src/engine/modelmanagement/factories/Factory.cpp:
```diff
@@ -95,6 +103,18 @@ void Factory::setAttributes(const YAML::Node& node, alica::AlicaElement* ael)
+        // Replace the placeholder with the corresponding abstract plan from the mapping
+        auto& mappings = Factory::modelManager->_planRepository._placeholderMappings;
+        for (auto& [mappingId, mapping] : mappings) {
+            if (auto abstractPlan = mapping->find(id); abstractPlan) {
```
minor: `auto *`

ntfshard replied:
and also put const around

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -36,13 +40,22 @@ ModelManager::ModelManager(ConfigChangeListener& configChangeListener, const std
     }
 
+    std::string placeholderMappingPath;
+    if (placeholderMapping.has_value()) {
```
empty filename is a valid name?

veerajsk replied:
this is changed to be a json now instead of a file name, therefore I think have an optional is useful

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -36,13 +40,22 @@ ModelManager::ModelManager(ConfigChangeListener& configChangeListener, const std
         AlicaEngine::abort(LOGNAME, "Cannot find MasterPlan '", masterPlanName, "'");
     }
 
+    std::string placeholderMappingPath;
```
variable scope can be reduced

veerajsk replied:
its removed now

veerajsk committed:
Take the mapping as a string instead of a file

Dale-Koenig reviewed (commented):
So rather than a fake use of placeholders, I was actually thinking we would use placeholders for the WaitForTrigger and WaitForMsg behaviors, so that we are using them in a natural way.  We allow loading from multiple folders, but ideally this always uses placeholders so that each folder is self contained and can be properly handled in alica web designer.  In this case the mapping one be in a config file in alica ros1 turtlesim package, and could be loaded with rosparam

veerajsk commented:
> So rather than a fake use of placeholders, I was actually thinking we would use placeholders for the WaitForTrigger and WaitForMsg behaviors, so that we are using them in a natural way. We allow loading from multiple folders, but ideally this always uses placeholders so that each folder is self contained and can be properly handled in alica web designer. In this case the mapping one be in a config file in alica ros1 turtlesim package, and could be loaded with rosparam

yes, this is a better use of placeholders in the current turtlesim. I'll make the changes (I think the current way of using placeholders was more useful for testing)

veerajsk committed:
Review comments

veerajsk committed:
Revert turtlesim state machine changes

veerajsk committed:
Use placeholders for wait for msg/trigger

Instead of having a new plan for placeholders, use the existing plans
& behaviours but use placeholders for WaitForMsg & WaitForTrigger
behaviours since these behaviours have different implementations for
ros1 & ros2 & therefore they serve the use case already

veerajsk committed:
Add placeholder test plan

veerajsk committed:
Placeholder alica test implementation

ntfshard commented on supplementary/alica_ros1/alica_ros_turtlesim/src/base_node.cpp:
```diff
@@ -49,7 +53,19 @@ int main(int argc, char** argv)
     killMyTurtle(name, privNh);
 
-    turtlesim::Base base(nh, privNh, name, agentId, roleset, masterPlan, {turtlesimRosPath, turtlesimLibPath});
+    std::optional<std::string> maybeMapping;
+    if (!placeholderMappingFilePath.empty()) {
+        std::ifstream mappingFileStream(placeholderMappingFilePath);
+        if (mappingFileStream) {
+            std::ostringstream ss;
+            ss << mappingFileStream.rdbuf();
+            maybeMapping.emplace(ss.str());
+        } else {
+            ROS_ERROR_STREAM("Could not read the placeholder mapping file");
+            return 1;
+        }
+    }
```
this code looks very similar to new function in alica_tests/include/test_alica.h

ntfshard commented on alica_engine/src/engine/modelmanagement/factories/Factory.cpp:
```diff
@@ -95,6 +102,18 @@ void Factory::setAttributes(const YAML::Node& node, alica::AlicaElement* ael)
+        // Replace the placeholder with the corresponding abstract plan from the mapping
+        const auto& mappings = Factory::modelManager->_planRepository._placeholderMappings;
+        for (const auto& [mappingId, mapping] : mappings) {
+            if (const auto abstractPlan = mapping->find(id); abstractPlan) {
```
minor: 
```suggestion
            if (const auto abstractPlan = mapping->find(id)) {
```

ntfshard commented on supplementary/alica_ros1/alica_ros_turtlesim/src/base.cpp:
```diff
@@ -20,11 +20,11 @@ namespace turtlesim
 {
     // Initialize Alica
-    ac = new alica::AlicaContext(AlicaContextParams(name, paths, roleset, master_plan, false, agent_id));
+    ac = new alica::AlicaContext(AlicaContextParams(name, paths, roleset, master_plan, false, agent_id, placeholderMapping));
```
minor: if we are already changing this line, maybe it's worth to move to smart pointers?

ntfshard reviewed (commented):
I don't have strong objections, but this PR is quite big to spot possible problems

bjoernschroeder committed:
change file extension from placeholders to .plh, rename attribute to isInterface

bjoernschroeder committed:
add placeholder default value to turtlesim test

bjoernschroeder committed:
use isInterface

bjoernschroeder committed:
use unique ids for blackboards

Dale-Koenig commented on supplementary/alica_ros1/alica_ros_turtlesim/etc/interfaces/WaitForMsg.beh:
```diff
@@ -19,9 +19,10 @@
   "eventDriven": false,
   "frequency": 5,
-  "id": 2324679431617631864,
+  "id": 1111111111111111112,
```
Why are these in a separate folder?  Is this how they are exported?
I would expect them still to just go into plans/behaviors folder as the only difference is the isInterface flag

bjoernschroeder replied:
this was probably done by hand, I changed it and made a reexport, they will be in behaviours folder

bjoernschroeder committed:
check correct mapping during initialization

bjoernschroeder committed:
format

bjoernschroeder committed:
reexport supplementary tests

bjoernschroeder committed:
reexport turtlesim interfaces

bjoernschroeder committed:
reexport plans for turtlesim base

bjoernschroeder committed:
use unique ptr for context

bjoernschroeder committed:
update supplementary tests

bjoernschroeder committed:
use copy of node

bjoernschroeder committed:
fix typo

Dale-Koenig committed:
Merge branch 'devel' into vsk_placeholders
